### PR TITLE
If statement for Iulia / Inline Assembly

### PR DESCRIPTION
**Pull request by chriseth**
This will be much more useful in the ABI decoder (only the encoder is part of develop as of now).

### CODE REVIEW

File: libjulia/backends/evm/EVMCodeTransform.cpp [hunk 1]
```diff
@@ -217,6 +217,18 @@ void CodeTransform::operator()(assembly::Instruction const& _instruction)
...
 }
 
+void CodeTransform::operator()(If const& _if)
+{
+	visitExpression(*_if.condition);
+	m_assembly.appendInstruction(solidity::Instruction::ISZERO);
+	AbstractAssembly::LabelID end = m_assembly.newLabelId();
+	m_assembly.appendJumpToIf(end);
+	(*this)(_if.body);
+	m_assembly.setSourceLocation(_if.location);
```
Comment (axic) [inline libjulia/backends/evm/EVMCodeTransform.cpp:228]: Shouldn't the sourcelocation be before the body or even the condition?

**Comment by axic** [comment]
I think this is missing the evm1.5 codegen part?

File: libjulia/backends/evm/EVMCodeTransform.cpp (hunk 1 above)
Comment (chriseth) [reply libjulia/backends/evm/EVMCodeTransform.cpp:228]: Hm, the source location of the label is correct, but what is wrong is the source location of the ISZERO and the jump above.

**Comment by chriseth** [comment]
No, AbstractAssembly handles both of them.

**Comment by axic** [comment]
> No, AbstractAssembly handles both of them.

Never mind, it is `EVMCodeTransform` and the if code has no need for evm15 specific stuff.

File: libjulia/backends/evm/EVMCodeTransform.cpp (hunk 1 above)
Comment (axic) [reply libjulia/backends/evm/EVMCodeTransform.cpp:228]: Will you look into this?

**Review by axic — CHANGES_REQUESTED** [review]
Please include the if statement in the formal spec

**Comment by axic** [comment]
It is lacking tests:
- printing test
- assemble test
- scope tests (`if iszero() { let x := 1 } x := 2`)
- tests in the julia suite

Though it seems `for` loop isn't covered in the printing test either.

File: libjulia/backends/evm/EVMCodeTransform.cpp (hunk 1 above)
Comment (axic) [reply libjulia/backends/evm/EVMCodeTransform.cpp:228]: Done.

**Comment by chriseth** [comment]
There is and end-to-end test for if - do we have an "assemble test suite"?

**Comment by axic** [comment]
`BOOST_AUTO_TEST_SUITE(Analysis)` in `test/solidity/InlineAssembly`

**Comment by chriseth** [comment]
Tests added.

File: test/libjulia/Parser.cpp [hunk 2]
```diff
@@ -269,6 +269,19 @@ BOOST_AUTO_TEST_CASE(multiple_assignment)
 	BOOST_CHECK(successParse(text));
 }
 
+BOOST_AUTO_TEST_CASE(if_statement)
+{
+	BOOST_CHECK(successParse("{ if 42:u256 {} }"));
```
Comment (axic) [inline test/libjulia/Parser.cpp]: Hmm, shouldn't in Julia this operate on boolean? `if true:bool {}`. It currently has a hidden behaviour of `if iszero(cond) {}`

File: docs/julia.rst [hunk 3]
```diff
@@ -248,8 +250,14 @@ We will use a destructuring notation for the AST nodes.
         G, L, break
     E(G, L, continue: BreakContinue) =
         G, L, continue
+    E(G, L, <if condition body>: If) =
+        let G0, L0, v = E(G, L, condition)
+        if v is true or non-zero:
```
Comment (axic) [inline docs/julia.rst]: This part is about julia, I think that is more strict and only considers `if v is true`.

File: test/libjulia/Parser.cpp (hunk 2 above)
Comment (chriseth) [reply test/libjulia/Parser.cpp]: I can change the test, but we do not yet enforce types anyway, right?

**Comment by chriseth** [comment]
Updated.

File: test/libjulia/Parser.cpp (hunk 2 above)
Comment (axic) [reply test/libjulia/Parser.cpp]: No, not yet. It is fine to leave as it is, we'll at least see if the enforcement triggers this failure :)

Comment (axic) [reply test/libjulia/Parser.cpp]: I see you've updated it, it is fine that way too.

**Comment by axic** [comment]
Perhaps we're failing zeppelin tests now as they've checking for even warnings:
```
> zeppelin-solidity@1.3.0 test /tmp/tmp.Cblq5apBSn
> scripts/test.sh
Starting our own testrpc instance
Error parsing /tmp/tmp.Cblq5apBSn/contracts/Bounty.sol: Warning: This is a pre-release compiler version, please do not use it in production.
Compilation failed. See above.
```

**Comment by chriseth** [comment]
@axic hm, I fear we have to patch that out, we cannot wait for modified zeppelin code when we introduce a new warning.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
Test failures are due to the Zeppelin issue. This should be fine to merge though as everything else works and it only adds the `if` feature.

**Pull request merged**